### PR TITLE
UGRID-22: SWIG C# wrapper

### PR DESCRIPTION
**Pull request by ahmad-el-sayed**

### CODE REVIEW

File: cmake/find_packages.cmake [hunk 1]
```diff
@@ -17,12 +17,21 @@ endif()
...
-find_package(hdf5 REQUIRED)
-if (hdf5_FOUND)
-    message(STATUS "Found HDF5 ${hdf5_VERSION}")
+find_package(HDF5)
+if (HDF5_FOUND)
+    message(STATUS "Found HDF5 ${HDF5_VERSION}")
+endif()
+
+# python
+set(Python_FIND_VIRTUALENV FIRST)
```
[the remark this answers is not in the document]
Comment (lucacarniato) [reply cmake/find_packages.cmake:31]: i see it is used later

File: libs/UGridNET/dll/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+add_library(
+  ${TARGET_NAME}
+  SHARED 
+  ${APP_CONFIG}
+  ${NUGET_CONFIG}
+  ${DIRECTORY_PACKAGES_PROPS}
+  ${DIRECTORY_BUILD_PROPS}
+  ${SWIG_GENERATED_CSHARP_SRCS}
```
Comment (lucacarniato) [inline libs/UGridNET/dll/CMakeLists.txt:56]: This variable is defined in the cmake project that runs before (add_subdirectory(SWIG)) and not used there. is it possible to define SWIG_GENERATED_CSHARP_SRCS  here?


file(GLOB SWIG_GENERATED_CSHARP_SRCS CONFIGURE_DEPENDS "${SWIG_OUTPUT_DIR}/*.cs")
set(SWIG_GENERATED_CSHARP_SRCS ${SWIG_GENERATED_CSHARP_SRCS} PARENT_SCOPE)

on the other hand, i understand that the swig project needs to run first

File: cmake/compiler_config.cmake [hunk 3]
```diff
@@ -11,19 +11,19 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 # Add compiler-specific options and definitions per supported platform
 if (UNIX)
   if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-    add_compile_options("-fvisibility=hidden;-Werror;-Wall;-Wextra;-pedantic;-Wno-unused-function")
-    add_compile_options("$<$<CONFIG:RELEASE>:-O2>")
-    add_compile_options("$<$<CONFIG:DEBUG>:-g>")
+    add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:-fvisibility=hidden;-Werror;-Wall;-Wextra;-pedantic;-Wno-unused-function>")
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/compiler_config.cmake:14]: The opposite. From gcc doc:
```
-Wunused-function
   Warn whenever a static function is declared but not defined or a non-inline
   static function is unused. This warning is enabled by ‘-Wall’.
```
`-Wno-unused-function` reverse that. notice the `no` . 

A cleanup in the future is necessary. Anything unused should be removed: classes, methods, variables, functions etc.

File: cmake/install_tools.cmake [hunk 4]
```diff
@@ -0,0 +1,114 @@
+if(WIN32)
+
+  function(check_arguments function_name args_prefix required_args)
+    set(missing_required_args FALSE)
+    set(missing_args_msg "")
+    foreach(arg IN LISTS required_args)
+      if (NOT DEFINED ${args_prefix}_${arg})
+        string(APPEND missing_args_msg "  ${arg} argument is required but it was not provided.\n")
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/install_tools.cmake:8]: Do you mean the length of the error message `missing_args_msg`? That's possible and more concise but less clear in my opinion.

File: cmake/install_tools.cmake [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    check_arguments("${CMAKE_CURRENT_FUNCTION}" "${prefix}" "${required_args}")
+
+    message(STATUS "Installing SWIG ${ARGS_SWIG_VERSION}...")
+
+    # Set the download URL and file paths
+    set(SWIG_URL "http://prdownloads.sourceforge.net/swig/swigwin-${ARGS_SWIG_VERSION}.zip")
+    set(DOWNLOAD_DIR "${ARGS_SWIG_INSTALL_PREFIX}/download")
+
+    # Create the download and extraction directories
+    file(MAKE_DIRECTORY ${ARGS_SWIG_INSTALL_PREFIX})
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/install_tools.cmake:44]: There is no need for the time being. This is done once, only in this config.

File: libs/UGridNET/dll/CMakeLists.txt (hunk 2 above)
Comment (ahmad-el-sayed) [reply libs/UGridNET/dll/CMakeLists.txt:56]: You can do file glob in one of either of the 2 configs. It does not matter. The path must be adapted. It makes sense though to create a list of the source in the config where they are generated.

File: cmake/install_tools.cmake [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    set(required_args ${options} ${one_value_args} ${multi_value_args})
+    check_arguments("${CMAKE_CURRENT_FUNCTION}" "${prefix}" "${required_args}")
+
+    message(STATUS "Installing NuGet ${ARGS_NUGET_VERSION}...")
+
+    set(NUGET_DOWNLOAD_DIR ${ARGS_NUGET_INSTALL_PREFIX}/nuget)
+    file(MAKE_DIRECTORY ${NUGET_DOWNLOAD_DIR})
+    set(NUGET_EXECUTABLE ${NUGET_DOWNLOAD_DIR}/nuget.exe)
+    if(NOT EXISTS ${NUGET_EXE})
+      set(NUGET_URL "https://dist.nuget.org/win-x86-commandline/${ARGS_NUGET_VERSION}/nuget.exe")
```
[the remark this answers is not in the document]
Comment (ahmad-el-sayed) [reply cmake/install_tools.cmake:103]: Similarities are minor: Only the download part and the messages.

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**